File: datastructures/LinkedList/LinkedList.java
package datastructures.LinkedList;
import java.io.*;

// Java program to implement a Singly Linked List
public class LinkedList
{
    LinkedListNode head;

    // Node structure for LinkedList
    static class LinkedListNode
    {
        int data;
        LinkedListNode next;

        public LinkedListNode(int nodeData)
        {
            this.data = nodeData;
            this.next = null;
        }
    }

    // Method to insert an element at the end of the LinkedList
    public static LinkedList insertNode(LinkedList list, int data)
    {
        LinkedListNode newNode = new LinkedListNode(data);
        newNode.next = null;
        LinkedListNode lastNode = list.head;

        // If the list is empty, set the new node as the head
        if(list.head == null)
        {
            list.head = newNode;
        } 
        else 
        {
            // Traverse to the last node and append the new node
            while(lastNode.next != null)
            {
                lastNode = lastNode.next;
            }
            lastNode.next = newNode;
        }
        return list;
    }

    // Method to insert an element at the beginning of the LinkedList
    public static LinkedList insertNodeAtFirst(LinkedList list, int data)
    {
        LinkedListNode newNode = new LinkedListNode(data);
        
        // If the list is empty, set the new node as the head
        if(list.head == null)
        {
            list.head = newNode;
        } 
        else
        {
            // Set the new node's next to the current head and update the head
            newNode.next = list.head;
            list.head = newNode;
        }
        return list;
    }

    // Method to delete an element at a specific position
    public static LinkedList deleteNode(LinkedList list, int pos)
    {
        LinkedListNode currNode = list.head;
        LinkedListNode prev = null;

        // If the position is 0 and the list is not empty, delete the head node
        if(pos == 0 && currNode != null)
        {
            list.head = currNode.next;
            System.out.println("Element at position " + pos + " deleted");
            return list;
        }
        
        int counter = 0;
        
        // Traverse to the specified position
        while(currNode != null)
        {
            if(counter == pos)
            {
                // Delete the node at the specified position by updating the previous node's next
                prev.next = currNode.next;
                System.out.println("Element at position " + pos + " deleted");
                break;
            }
            else
            {
                counter++;
                prev = currNode;
                currNode = currNode.next;
            }
        }
        
        // If the specified position is not found
        if(currNode == null)
        {
            System.out.println("Element at position " + pos + " not found");
        }
        return list;
    }

    // Method to print the LinkedList
    public static void printList(LinkedList list)
    {
        LinkedListNode currNode = list.head;

        while(currNode != null)
        {
            System.out.print(currNode.data + " ");
            currNode = currNode.next;
        }
    }

    public static void main(String[] args) 
    {
        LinkedList list = new LinkedList();

        // Insert nodes into the LinkedList
        insertNode(list, 0);
        insertNode(list, 2);
        insertNodeAtFirst(list, 4);
        insertNode(list, 5);
        insertNodeAtFirst(list, 8);

        // Print the LinkedList
        printList(list);
    }
}

File: datastructures/LinkedList/LinkedList.c
#include "LinkedList.h"

/**
 * @brief Inserts a new node at the beginning of the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void insertNodeAtBeginning(struct Node **List)
{
    // Allocate memory for the new node
    struct Node *tempNode = NULL;
    tempNode = (struct Node *) malloc(sizeof(struct Node));

    if(NULL != tempNode)
    {
        // Prompt user for input
        printf("Enter Node Data : ");
        scanf("%i", &tempNode->NodeData);

        if(NULL == *List)
        {
            tempNode->NodeLink =NULL;
            *List = tempNode;
        }
        else
        {
            tempNode->NodeLink = *List;
            *List = tempNode;
        }
    }
    else
    {
        printf("Out of Memory Space:\n");
    }
}

/**
 * @brief Inserts a new node at the end of the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void insertNodeAtEnd(struct Node **List)
{
    // Allocate memory for the new node
    struct Node *tempNode = NULL;
    struct Node *lastNode = NULL;
    tempNode = (struct Node *) malloc(sizeof(struct Node));

    if(NULL != tempNode)
    {
        // Prompt user for input
        printf("Enter Node Data : ");
        scanf("%i", &tempNode->NodeData);
        tempNode->NodeLink = NULL;

        if(NULL == *List)
        {
            *List = tempNode;
        }
        else
        {   
            lastNode = *List;
            while (lastNode->NodeLink != NULL)
            {
                lastNode = lastNode->NodeLink;
            }
            lastNode->NodeLink = tempNode;
        }
    }
    else
    {
        printf("Out of Memory Space:\n");
    }
}

/**
 * @brief Inserts a new node after a specified position in the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void insertNodeAfter(struct Node *List)
{
    struct Node *tempNode = NULL, *nodeListCounter = List;
    uint32 nodePosition = ZERO;    /*  */
    uint32 listLength = ZERO;      /*  */
    uint32 nodeCounter = 1;     /* Points to the first node */

    // Prompt user for position
    printf("Enter Node Position (First node is position = 1): ");
    scanf("%i", &nodePosition);

    listLength = getLength(List);

    if(nodePosition > listLength)
    {
        printf("Invalid Node Position -> List has ( %i ) Nodes \n", listLength);
    }
    else
    {
        // Find the node after which to insert
        for (nodeCounter = 1; nodeCounter < nodePosition; nodeCounter++)
        {
            nodeListCounter = nodeListCounter->NodeLink;
        }

        tempNode = (struct Node *)malloc(sizeof(struct Node));
        if(NULL != tempNode)
        {
            // Prompt user for input
            printf("Enter Node Data : ");
            scanf("%i", &tempNode->NodeData);

            // Insert the new node
            tempNode->NodeLink = nodeListCounter->NodeLink;
            nodeListCounter->NodeLink = tempNode;
        }
    }
}

/**
 * @brief Deletes the first node in the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void deleteNodeAtBeginning(struct Node**List)
{
    struct Node *tempNode = *List;
    uint32 listLength = ZERO;

    listLength = getLength(*List);
    if (ZERO == listLength)
    {
        printf("List is Empty , nothing to be deleted !! \n");
    }
    else
    {
       // Update the head and free the first node
       *List = tempNode->NodeLink;
       tempNode->NodeLink = NULL;
       free(tempNode); 
    }
}

/**
 * @brief Deletes a node at a specified position in the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void deleteNode(struct Node **List)
{
    struct Node *nextNode = NULL;
    struct Node *nodeListCounter = *List;
    uint32 nodePosition = ZERO;
    uint32 listLength = ZERO;
    uint32 nodeCounter = 1;

    // Prompt user for position
    printf("Enter Node Position (First node is position = 1): ");
    scanf("%i", &nodePosition);

    listLength = getLength(*List);

    if (nodePosition > listLength)
    {
        printf("Invalid Node Position -> List has ( %i ) Nodes\n", listLength);
    }
    else if (1 == listLength)
    {
        deleteNodeAtBeginning(List);
    }
    else
    {
        // Traverse to the node before the one to be deleted
        for (nodeCounter = 1; nodeCounter < (nodePosition - 1); nodeCounter++)
        {
            nodeListCounter = nodeListCounter->NodeLink;
        }
        nextNode = nodeListCounter->NodeLink;

        // Update pointers and free the deleted node
        nodeListCounter->NodeLink = nextNode->NodeLink;
        free(nextNode);
    }
}

/**
 * @brief Displays all nodes in the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 */
void displayAllNodes(struct Node *List)
{
    struct Node *tempNode = List;

    printf("Node Values : ");
    if(NULL == tempNode)
    {
        printf("List is empty !! \n");
    }
    else
    {
        while(tempNode != NULL)
        {
            printf("%i -> ", tempNode->NodeData);
            tempNode = tempNode->NodeLink;
        }
        if(NULL == tempNode)
        {
            printf("NULL \n");
        }
    }
}

/**
 * @brief Calculates the length of the linked list.
 * 
 * @param List A pointer to the head of the linked list.
 * @return uint32 The length of the linked list.
 */
uint32 getLength(struct Node *List)
{
    struct Node *tempNode = List;
    uint32 nodeCount = ZERO;

    while(tempNode != NULL)
    {
        nodeCount++;
        tempNode = tempNode->NodeLink;
    }

    return nodeCount;
}
